File: internal/jsonc/ast/ast.go
package ast

type NodeType uint8

const (
	NodeTypeUnknown NodeType = iota
	NodeTypeObject
	NodeTypeArray
	NodeTypeValue
	NodeTypeDoc

	LabelNodeTypeUnknown = "unknown"
	LabelNodeTypeObject  = "object"
	LabelNodeTypeArray   = "array"
	LabelNodeTypeValue   = "value"
	LabelNodeTypeDoc     = "doc"
)

func (nt NodeType) String() string {
	switch nt {
	case NodeTypeObject:
		return LabelNodeTypeObject
	case NodeTypeArray:
		return LabelNodeTypeArray
	case NodeTypeValue:
		return LabelNodeTypeValue
	case NodeTypeDoc:
		return LabelNodeTypeDoc
	default:
		return LabelNodeTypeUnknown
	}
}

func ParseNodeType(s string) NodeType {
	switch s {
	case LabelNodeTypeObject:
		return NodeTypeObject
	case LabelNodeTypeArray:
		return NodeTypeArray
	case LabelNodeTypeValue:
		return NodeTypeValue
	case LabelNodeTypeDoc:
		return NodeTypeDoc
	default:
		return NodeTypeUnknown
	}
}

type Node interface {
	GetTag() *Tag
	GetType() NodeType

	GetPath() string
	SetPath(path string)
}

type Field struct {
	Key   string
	Value Node
}

type Object struct {
	Fields []*Field
	Tag    *Tag
	Path   string
}

func (o *Object) GetPath() string { return o.Path }

func (o *Object) SetPath(path string) { o.Path = path }

func (o *Object) GetType() NodeType { return NodeTypeObject }

func (o *Object) GetTag() *Tag {
	if o == nil {
		return nil
	}
	if o.Tag != nil {
		return o.Tag
	}
	return nil
}

type Array struct {
	Items []Node
	Tag   *Tag
	Path  string
}

func (a *Array) GetPath() string { return a.Path }

func (a *Array) SetPath(path string) { a.Path = path }

func (a *Array) GetType() NodeType { return NodeTypeArray }

func (a *Array) GetTag() *Tag {
	if a == nil {
		return nil
	}
	if a.Tag != nil {
		return a.Tag
	}
	return nil
}

type Value struct {
	Data any
	Text string
	Tag  *Tag
	Path string
}

func (v *Value) GetPath() string { return v.Path }

func (v *Value) SetPath(path string) { v.Path = path }

func (v *Value) GetType() NodeType { return NodeTypeValue }

func (v *Value) GetTag() *Tag {
	if v == nil {
		return nil
	}
	if v.Tag != nil {
		return v.Tag
	}
	return nil
}

type Doc struct {
	Fields []*Field
	Tag    *Tag
	Path   string
}

func (d *Doc) GetPath() string { return d.Path }

func (d *Doc) SetPath(path string) { d.Path = path }

func (d *Doc) GetType() NodeType { return NodeTypeDoc }

func (d *Doc) GetTag() *Tag {
	if d == nil {
		return nil
	}
	if d.Tag != nil {
		return d.Tag
	}
	return nil
}

File: mm-swift/Sources/MetaMessage/Ir/AST.swift
import Foundation

public enum NodeType: String {
    case unknown
    case object
    case array
    case value
    case doc
}

public protocol Node {
    func getTag() -> Tag?
    func getType() -> NodeType
    func getPath() -> String
    func setPath(_ path: String)
    func setTag(_ tag: Tag)
}

public struct Field {
    public var key: String
    public var value: Node

    public init(key: String, value: Node) {
        self.key = key
        self.value = value
    }
}

public class MMObject: Node {
    public var fields: [Field]
    public var tag: Tag?
    public var path: String

    public init(fields: [Field] = [], tag: Tag? = nil, path: String = "") {
        self.fields = fields
        self.tag = tag
        self.path = path
    }

    public func getTag() -> Tag? {
        return tag
    }

    public func getType() -> NodeType {
        return .object
    }

    public func getPath() -> String {
        return path
    }

    public func setPath(_ path: String) {
        self.path = path
    }

    public func setTag(_ tag: Tag) {
        self.tag = tag
    }
}

public class MMArray: Node {
    public var items: [Node]
    public var tag: Tag?
    public var path: String

    public init(items: [Node] = [], tag: Tag? = nil, path: String = "") {
        self.items = items
        self.tag = tag
        self.path = path
    }

    public func getTag() -> Tag? {
        return tag
    }

    public func getType() -> NodeType {
        return .array
    }

    public func getPath() -> String {
        return path
    }

    public func setPath(_ path: String) {
        self.path = path
    }

    public func setTag(_ tag: Tag) {
        self.tag = tag
    }
}

public class Value: Node {
    public var data: Any?
    public var text: String
    public var tag: Tag?
    public var path: String

    public init(data: Any? = nil, text: String = "", tag: Tag? = nil, path: String = "") {
        self.data = data
        self.text = text
        self.tag = tag
        self.path = path
    }

    public func getTag() -> Tag? {
        return tag
    }

    public func getType() -> NodeType {
        return .value
    }

    public func getPath() -> String {
        return path
    }

    public func setPath(_ path: String) {
        self.path = path
    }

    public func setTag(_ tag: Tag) {
        self.tag = tag
    }
}

public class MMDoc: Node {
    public var fields: [Field]
    public var tag: Tag?
    public var path: String

    public init(fields: [Field] = [], tag: Tag? = nil, path: String = "") {
        self.fields = fields
        self.tag = tag
        self.path = path
    }

    public func getTag() -> Tag? {
        return tag
    }

    public func getType() -> NodeType {
        return .doc
    }

    public func getPath() -> String {
        return path
    }

    public func setPath(_ path: String) {
        self.path = path
    }

    public func setTag(_ tag: Tag) {
        self.tag = tag
    }
}